### PR TITLE
Added templateId field in google_data_loss_prevention_deidentify_template and google_data_loss_prevention_inspect_template

### PR DESCRIPTION
**Pull request by abheda-crest**
Added `template_id` field to `google_data_loss_prevention_deidentify_template` and `google_data_loss_prevention_inspect_template` resource.
fixes https://github.com/hashicorp/terraform-provider-google/issues/13381

If this PR is for Terraform, I acknowledge that I have:

- [x] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [x] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/third_party/terraform/tests) (for handwritten resources or update tests).
- [x] [Generated Terraform providers](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/docs/content/docs/getting-started/generate-providers.md), and ran [`make test` and `make lint`](https://googlecloudplatform.github.io/magic-modules/docs/getting-started/run-provider-tests/#run-unit-tests) in the generated providers to ensure it passes unit and linter tests.
- [x] [Ran](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/docs/content/docs/getting-started/run-provider-tests.md) relevant acceptance tests using my own Google Cloud project and credentials (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [x] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
dlp: added `template_id` field to `google_data_loss_prevention_deidentify_template` and `google_data_loss_prevention_inspect_template`
```

### CODE REVIEW

**Comment by abheda-crest** [comment]
Here, I've made the `template_id` fields as `url_param_only` to avoid generating the flattener and expander functions for the same. We need to skip the expand function because the existing resource [wraps the fields in terraform](https://github.com/hashicorp/terraform-provider-google/blob/main/google/resource_data_loss_prevention_deidentify_template.go#L10353). e.g., the `deidentifyTemplate` field in the case of the `deidentify_template` resource. `templateId` is in the same hierarchy as this [wrapped object](https://cloud.google.com/dlp/docs/reference/rest/v2/projects.deidentifyTemplates/create#request-body). Similarly, we need to skip the flatten function as the [GET endpoint](https://cloud.google.com/dlp/docs/reference/rest/v2/projects.inspectTemplates#resource:-inspecttemplate) doesn't contain the `templateId` field as is; instead, we need to extract it from the `name` output-only field. The response from the GET endpoint is the same wrapped object that we passed while creating the resource.

**Comment by abheda-crest** [comment]
@roaks3 Could you please start the tests as it is awaiting your approval?

File: mmv1/products/dlp/DeidentifyTemplate.yaml
Comment (roaks3) [inline mmv1/products/dlp/DeidentifyTemplate.yaml:49]: The more common way to do this is choose any name with a hyphen, eg. `my-template`. This will show up as `my-template` in the docs, and `tf-test-my-template%{random}` in the test. This also means you can get rid of the `test_vars_overrides` block.

File: mmv1/products/dlp/InspectTemplate.yaml
Comment (roaks3) [inline mmv1/products/dlp/InspectTemplate.yaml:52]: Same comment as above

**Review by roaks3 — COMMENTED** [review]
I'll plan to take a second pass on this, but a few high-level comments:
- This may be a little easier to split into two changesets: one to include the field on one of these resources, and then the second would build on that by reusing some of the same functions. Up to you, but there are tradeoffs, and this PR might take a little longer as it is.
- I think we want to test an update to these fields
- I'm wondering if there are any existing patterns we can reuse regarding the custom encoders/decoders (I'll check up on that)

File: mmv1/products/dlp/DeidentifyTemplate.yaml
Comment (abheda-crest) [reply mmv1/products/dlp/DeidentifyTemplate.yaml]: Thanks for pointing this out. I've updated this.

File: mmv1/products/dlp/InspectTemplate.yaml
Comment (abheda-crest) [reply mmv1/products/dlp/InspectTemplate.yaml]: I've updated the same.

**Comment by abheda-crest** [comment]
> I'll plan to take a second pass on this, but a few high-level comments:
> 
> * This may be a little easier to split into two changesets: one to include the field on one of these resources, and then the second would build on that by reusing some of the same functions. Up to you, but there are tradeoffs, and this PR might take a little longer as it is.
> * I think we want to test an update to these fields

I've pushed commits that address these comments as follows:
* I've merged the separate decoders for these resources into a [single decoder](https://github.com/GoogleCloudPlatform/magic-modules/pull/7986/files#diff-3d1d210b1a91ea8ecd14c10ff15d9e9db53e21a331244956bc1daac570d9c4f7). It seems like we will likely need shared encoder, decoder, and update encoder for these resources. Though your input will be helpful for the same.
* I've added update tests for `template_id` field.

**Comment by abheda-crest** [comment]
@roaks3 Could you please start the tests as it is awaiting your approval?

**Comment by abheda-crest** [comment]
Hi @roaks3, as the tests have passed now, could you please review this PR? Thanks!

**Comment by roaks3** [comment]
@abheda-crest Great, thanks for adding those tests! Everything looks good at a high level, but I'm going to take one last look through tomorrow morning to be sure, but I expect this is likely ready to be merged.

File: mmv1/products/dlp/DeidentifyTemplate.yaml
Comment (roaks3) [inline mmv1/products/dlp/DeidentifyTemplate.yaml:55]: I think it might be easier to reuse `templates/terraform/encoders/wrap_object.go.erb`, instead of duplicating that functionality in the `update_encoder` directory. Would that work here? (Same for InspectTemplate)

And, just to verify, I assume that `wrap_object_with_template_id` doesn't work for this case? (presumably because `templateId` cannot be updated?)

File: mmv1/third_party/terraform/tests/resource_data_loss_prevention_deidentify_template_test.go
Comment (roaks3) [inline mmv1/third_party/terraform/tests/resource_data_loss_prevention_deidentify_template_test.go:1113]: This is 100% my fault for requesting the update tests, but I think we don't need them here, because the field is `immutable` (I missed that on the first look).

Very sorry for the thrash, but I would remove both sets of handwritten `*Update` tests (in the future, if this were an updatable field, it is typically best to include a new field in the existing update test, when one exists).

File: mmv1/products/dlp/DeidentifyTemplate.yaml
Comment (roaks3) [inline mmv1/products/dlp/DeidentifyTemplate.yaml:46]: Looking more closely at the behavior for this field, I would reduce the examples you've included:
- Both examples are very redundant with the existing `basic` examples, and I don't think they provide real value in the docs, so I would include `skip_docs: true` for both.
- For that same reason, it is probably not necessary to test all of the same fields, so ideally we could remove all but one of the `transformations`, and also remove optional fields like `description` and `display_name`.

**Review by roaks3 — COMMENTED** [review]
Functionality LGTM, just had a couple of comments on the tests/examples. I'll wait to hear your thoughts, but this is otherwise ready for merge.

File: mmv1/products/dlp/DeidentifyTemplate.yaml
Comment (abheda-crest) [reply mmv1/products/dlp/DeidentifyTemplate.yaml]: Yeah, Reusing the `templates/terraform/encoders/wrap_object.go.erb` would definitely work. Also, while updating the resource, the [request body](https://cloud.google.com/dlp/docs/reference/rest/v2/projects.deidentifyTemplates/patch#request-body) doesn't expect `templateId` field. Hence, I've used two different encoders for creating and updating the template.

File: mmv1/products/dlp/DeidentifyTemplate.yaml
Comment (abheda-crest) [reply mmv1/products/dlp/DeidentifyTemplate.yaml:46]: Sure. I've added `skip_docs: true` and removed redundant fields from both of these examples.

**Comment by abheda-crest** [comment]
@roaks3 Do you have any idea why the `presubmit-generate-diffs` are failing?

**Comment by roaks3** [comment]
Yes, the `presubmit-generate-diffs` failure was on our end yesterday (affected all PRs), and has since been resolved. I just kicked off the tests one last time, and will approve/merge if they look good.

**Review by roaks3 — APPROVED** [review]

**Pull request merged**
